packages/theme/src/cli/utilities/theme-uploader.ts

```diff
@@ -75,27 +77,85 @@ export function uploadTheme(
     renderThemeSyncProgress: async () => {
       if (options?.deferPartialWork) return
 
-      const {progress: uploadProgress, promise: uploadPromise} = await uploadJobPromise
-      await renderTasksToStdErr(
-        createIntervalTask({
-          promise: uploadPromise,
-          titleGetter: () => `Uploading files to remote theme ${getProgress(uploadProgress)}`,
-          timeout: 1000,
-        }),
-      )
-
-      const {progress: deleteProgress, promise: deletePromise} = await deleteJobPromise
-      await renderTasksToStdErr(
-        createIntervalTask({
-          promise: deletePromise,
-          titleGetter: () => `Cleaning your remote theme ${getProgress(deleteProgress)}`,
-          timeout: 1000,
-        }),
-      )
+      if (options?.multiEnvironment) {
+        await renderMultiEnvironmentProgress(uploadJobPromise, deleteJobPromise, options, getProgress)
+      } else {
+        const {progress: uploadProgress, promise: uploadPromise} = await uploadJobPromise
+
+        await renderTasksToStdErr(
+          createIntervalTask({
+            promise: uploadPromise,
+            titleGetter: () => `Uploading files to remote theme ${getProgress(uploadProgress)}`,
+            timeout: 1000,
+          }),
+        )
+
+        const {progress: deleteProgress, promise: deletePromise} = await deleteJobPromise
+        await renderTasksToStdErr(
+          createIntervalTask({
+            promise: deletePromise,
+            titleGetter: () => `Cleaning your remote theme ${getProgress(deleteProgress)}`,
+            timeout: 1000,
+          }),
+        )
+      }
+
+      // Report any failed uploads after all progress rendering is complete
+      reportFailedUploads(uploadResults, options?.environment)
     },
   }
 }
 
+/**
+ * Display multi environment upload progress with concurrency-friendly
+ * stderr.write output rather than the animated progress bar
+ */
+async function renderMultiEnvironmentProgress(
+  uploadJobPromise: Promise<SyncJob>,
+  deleteJobPromise: Promise<SyncJob>,
+  options: UploadOptions,
+  getProgress: (params: {current: number; total: number}) => string,
+) {
+  const write = (message: string) =>
+    process.stderr.write(`${timestampDateFormat.format(new Date())} │ ${options.environment} │ ${message}\n`)
```

Comment:
The output format is based off `app dev`:

<img width="710" height="225" alt="app-output" src="https://github.com/user-attachments/assets/4592b629-2900-44f3-a5ea-f263b846f3bb" />